### PR TITLE
Implement sync for credentials

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/488551667048375/1205082906974103/f 

### Description
Allows user to sync their credentials across devices.

### Steps to test this PR
see notes in https://app.asana.com/0/488551667048375/1205082906974103/f

### CODE REVIEW

**Comment by cmonfortep** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3366** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3366" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3366?utm_source=stack-comment).

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt
Comment (CDRussell) [inline autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt:93]: I'm assuming this `modifiedSince` is milliseconds since 1970, but would be good to make that more explicit here (e.g, add javadoc for the parameter)

Comment (CDRussell) [inline autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt:93]: Since changing a public API, should be following this right?: [[Doc] Android Public APIs](https://app.asana.com/0/1202561462274611/1205060262189411/f)

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/SecureStoreBackedAutofillStore.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/SecureStoreBackedAutofillStore.kt:189]: Probably related to https://app.asana.com/0/0/1205110657412668/f

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncDataPersister.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncDataPersister.kt:85]: typo: 
`crendentials` used in many places in this class instead of `credentials`

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt:34]: it's nice where you have used `syncId` to make it explicit what kind of ID it is; being more explicit that the `id` is a `localId` might also be good in this class

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt:60]: is `insert` right here? maybe i'm missing the greater context, but seems confusing to call `insert` from a function dealing with deletions

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:39]: do we have to use `runBlocking` here? which thread are we blocking?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:37]: checking my understanding here: I guess this means if there is any error with any entry, we call the whole merge operation a failure? i.e, we don't want to ignore a failing entry and process the rest?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:35]: There's a lot going on in this method; finding it hard to process (pun intended) all that it's doing. Not sure if you could refactor out some logic into standalone functions perhaps to help break it down a bit.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsMergeStrategy.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsMergeStrategy.kt:23]: The API gives no indication that it will be expensive to run or block the calling thread internally. We should at least mark this with a worker thread annotation?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt:36]: Same comment as on the other merge strategy; having everything in one large function is hurting readability of the code IMO

**Review by CDRussell — COMMENTED** [review]
Still working through this review but figure i'll flush out the pending comments so far for consideration

File: autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt
Comment (CDRussell) [inline autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt:35]: we should mention the format of the string (presumably ISO-8601?)

File: secure-storage/secure-storage-api/src/main/java/com/duckduckgo/securestorage/api/SecureStorage.kt
Comment (CDRussell) [inline secure-storage/secure-storage-api/src/main/java/com/duckduckgo/securestorage/api/SecureStorage.kt:99]: > If [canAccessSecureStorage] is false when this is invoked, an empty flow will be emitted.

if this does happen, what will sync do with that?

File: sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt
Comment (CDRussell) [inline sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt:21]: you should do it 😄

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/SyncApiClient.kt
Comment (CDRussell) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/SyncApiClient.kt:89]: kill this?

**Review by CDRussell — COMMENTED** [review]
.

File: autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt
Comment (cmonfortep) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt:93]: This is now removed as I'm moving away from using AutofillStore (as discussed)

Comment (cmonfortep) [reply autofill/autofill-api/src/main/java/com/duckduckgo/autofill/api/store/AutofillStore.kt:93]: This is now removed as I'm moving away from using AutofillStore (as discussed)

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/SecureStoreBackedAutofillStore.kt
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/SecureStoreBackedAutofillStore.kt:189]: all good here (when user deletes, we update delete timestamp in metadata db)

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt:34]: also updated at db level, sounds good

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncMetadata.kt:60]: discussed offline: it's updating `deleted_at` timestamp

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:39]: annotated the interface with WorkerThread. Strategies are used as part of Persister, which also checks we are not in main thread.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsMergeStrategy.kt
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsMergeStrategy.kt:23]: this is now annotated thanks.

File: secure-storage/secure-storage-api/src/main/java/com/duckduckgo/securestorage/api/SecureStorage.kt
Comment (cmonfortep) [reply secure-storage/secure-storage-api/src/main/java/com/duckduckgo/securestorage/api/SecureStorage.kt:99]: Since we have timestamp in metadata table, we don't need to use this method anymore 🙌

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt:52]: Nit: this could be a good use case for [inline value classes](https://kotlinlang.org/docs/inline-classes.html)

the string parameter could accept any string, but has a very specific format required in order for it to be used. could define something like

`value class Iso8601String(private val s: String)` that is passed into these methods instead of just a string. might help make it clearer what is expected and avoid future errors by forcing the caller to be a little bit more deliberate in the string that is passed

just a suggestion to consider

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt:138]: is there a better name than `values2`?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncStore.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSyncStore.kt:50]: not a big deal, but should we match the file's name package to the filename here?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/SyncDateProvider.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/SyncDateProvider.kt:24]: nit: it might make things easier to test if this was a interface we injected in DI, and could replace it with a double during testing. we do that elsewhere, e.g.

```
interface TimeProvider {
    fun currentTimeMillis(): Long
}

@ContributesBinding(AppScope::class)
class SystemCurrentTimeProvider @Inject constructor() : TimeProvider {
    override fun currentTimeMillis(): Long = System.currentTimeMillis()
}
```

 feel free to ignore if testing what happens at different times isn't a problem

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt:31]: Typo: class name should start with a capital letter

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt
Comment (CDRussell) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CrendentialsDedupStrategy.kt:42]: As per similar comment elsewhere, this code is hard to follow as is and could benefit from being pulled out into several smaller functions I reckon.

File: autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt
Comment (CDRussell) [reply autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt:48]: bump

File: sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt
Comment (CDRussell) [inline sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt:49]: do it?

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt
Comment (CDRussell) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt:114]: delete this?

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt
Comment (CDRussell) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt:167]: Won't comment them all, but these comments should be deleted / clear up somehow

File: sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncEngineTest.kt
Comment (CDRussell) [inline sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncEngineTest.kt:84]: private task

**Review by CDRussell — APPROVED** [review]
Approving as the various edge cases all seem resolved (🎉).

There's still a few comments from last time that aren't addressed, and a few more to consider. Approving assuming they will all be addressed (fix or comment why not) before merging.

Looking good!

File: sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncEngineTest.kt
Comment (cmonfortep) [reply sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncEngineTest.kt:84]: Public now 👍

File: sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt
Comment (cmonfortep) [reply sync/sync-api/src/main/java/com/duckduckgo/sync/api/engine/Models.kt:49]: That was for me to create a task in Asana, not to fix it as part of this PR.

Changed comment to refer to asana instead 👍

File: autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt
Comment (cmonfortep) [reply autofill/autofill-store/src/main/java/com/duckduckgo/autofill/store/CredentialsSyncMetadataDao.kt:48]: Documented that in the Entity class, also in the `CredentialsSyncMetadata` class. Dao should be used only internally.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/CredentialsSync.kt:138]: `values007` jk 

good catch, I can go for `changes` instead.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/SyncDateProvider.kt
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/SyncDateProvider.kt:24]: Just introduced this wrapper for clarity when reading the code, I'm good not injecting this for now.

**Pull request merged**
